### DISCUSSION TITLE
How to connect the VM that created with the command in test_snapshot_restore_hotplug_virtiomem

### DISCUSSION THREAD

**Discussion by peng6662001** [cloud-hypervisor/cloud-hypervisor#5274] [Q&A] [answered]
I run a command to test VM on Ampere Altra:
```
rust-gdb --args ./target/aarch64-unknown-linux-gnu/debug/deps/integration-4bccc5f1331cf928 common_parallel::test_snapshot_restore_hotplug_virtiomem
```
The command comes from a script:
```
./scripts/dev_cli.sh tests --integration -- --test-filter test_snapshot_restore_hotplug_virtiomem
```
The command create a VM with the command:
```
"target/aarch64-unknown-linux-gnu/release/cloud-hypervisor" "--api-socket" "/tmp/chgIrlCt/cloud-hypervisor.sock.1" "--event-monitor" "path=/tmp/chgIrlCt/event.json" "--cpus" "boot=4" "--memory" "size=4G,hotplug_m
ethod=virtio-mem,hotplug_size=32G" "--balloon" "size=0" "--kernel" "/root/workloads/vmlinux" "--disk" "path=/tmp/chgIrlCt/osdisk.img" "--disk" "path=/tmp/chgIrlCt/cloudinit,iommu=on" "--net" "id=net123,tap=,mac=
12:34:56:78:90:01,ip=192.168.1.1,mask=255.255.255.0" "--vsock" "cid=3,socket=/tmp/chgIrlCt/vsock" "--cmdline" "root=/dev/vda1 console=hvc0 rw systemd.journald.forward_to_console=1" "-v"
```

If I set a breakpoint at test_infra/src/lib.rs:113
```
let num_events = match epoll::wait(epoll_fd, timeout * 1000_i32, &mut events[..])
```
I can't connect the VM with "ssh cloud@192.168.1.2"

If I set a breakpoint at test_infra/src/lib.rs:114
```
Ok(num_events) => Ok(num_events)
```
The VM can be connected with ssh.

I can't understand it,why the VM can be connected after epoll::wait?How to config the IP of "192.168.1.2" for VM automatically.

**Comment by likebreath** [comment]
The "epoll" you mentioned above is essentially how the guest notifies the host (e.g. the integration test) that the guest is fully booted and properly configured for its network, after which the guest is ready to take ssh commands from the host. So with breakpoint at `at test_infra/src/lib.rs:113`, you will need to wait manually (say for another minute) before you can reach the guest via ssh.

We are using cloudinit to setup the guest with the notifying service as I mentioned above as well as the guest network configuration. Details can be found:

[1] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/test_data/cloud-init/ubuntu/user-data#L70-L76
[2] https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/test_data/cloud-init/ubuntu/network-config

The guest ip is assigned dynamically by the integration test. Please check details here: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/test_infra/src/lib.rs#L773-L779.

**Reply by peng6662001** [reply]
@likebreath I created a VM manually, and the login prompt is also displayed.

The terminal keeps printing some error logs:
```
[ 4024.275896] systemd[1]: notify-booted.service: Scheduled restart job, restart counter is at 1640.
[ 4024.276443] systemd[1]: Stopped Notify the tcp listener on the host that the guest is booted.
[ 4024.307836] systemd[1]: Started Notify the tcp listener on the host that the guest is booted.
[ 4024.311537] cloud-hypervisor-notify-booted.sh[3068]: /usr/bin/cloud-hypervisor-notify-booted.sh: line 4: @TCP_LISTENER_PORT: Servname not supported for ai_socktype
[ 4024.312108] cloud-hypervisor-notify-booted.sh[3068]: /usr/bin/cloud-hypervisor-notify-booted.sh: line 4: /dev/tcp/@HOST_IP/@TCP_LISTENER_PORT: Invalid argument
[ 4024.312619] systemd[1]: notify-booted.service: Main process exited, code=exited, status=1/FAILURE
[ 4024.313078] systemd[1]: notify-booted.service: Failed with result 'exit-code'.
```

I can't ping 192.168.1.2 on host.

Could you please tell me how to do with shell command

**Comment by peng6662001** [comment, marked as answer]
Modify test_data/cloud-init/ubuntu/user-data
```
-        echo -n "@DEFAULT_TCP_LISTENER_MESSAGE" > /dev/tcp/@HOST_IP/@TCP_LISTENER_PORT
+        echo -n "booted" > /dev/tcp/192.168.2.1/8001
```

Make test.sh:
```
#!/bin/bash -x

rm -rf /tmp/ubuntu-cloudinit.img
rm -rf /tmp/ch0
rm -rf /tmp/vsock

./scripts/create-cloud-init.sh

cp /root/workloads/focal-server-cloudimg-arm64-custom-20210929-0.raw /root/ch_files/osdisk.img
target/aarch64-unknown-linux-gnu/release/cloud-hypervisor \
        --api-socket /tmp/ch0 \
        --event-monitor path=/tmp/event.json \
        --cpus boot=4 \
        --memory size=4G,hotplug_method=virtio-mem,hotplug_size=32G \
        --balloon size=0 \
        --kernel /root/workloads/Image \
        --disk path=/root/ch_files/osdisk.img \
        --disk path=/tmp/ubuntu-cloudinit.img,iommu=on \
        --net id=net123,tap=,mac=12:34:56:78:90:ab,ip=192.168.2.1,mask=255.255.255.0 \
        --vsock cid=3,socket=/tmp/vsock \
        --cmdline "console=hvc0 root=/dev/vda1 rw systemd.journald.forward_to_console=1"
```
Run "nc -l 8001" on host,after receive "booted",I can connect the VM by `ssh cloud@192.168.2.2`
